src/pages/Checkout/Checkout.js

```diff
@@ -119,21 +109,24 @@ export class Checkout extends PureComponent {
       )
     );
 
-    const goToShopifyCheckout = () => {
-      client.checkout
-      // update checkoutId to ensure that a fresh checkout is used
-      // this avoids the edge case where a user checks out but doesn't finish
-      // and then returns to the store to update their cart and re-checkout
-        .create().then(checkout => {
-          updateCheckoutId(checkout.id)
-        })
-        .then(client.checkout.addLineItems(checkoutId, lineItemsToAdd)
-        .then(checkout => {
-          window.location.href = checkout.webUrl
-      }))
-    };  
+   const createIdAndCheckout = () =>
+    client.checkout.create()
+      .then(checkout => {
+        updateCheckoutId(checkout.id)
+        return checkout.id
+      })
+      .then(checkoutId =>
+        client.checkout.addLineItems(checkoutId, lineItemsToAdd)
+          .then(checkout => {
+            window.location.href = checkout.webUrl;
+          })
+          .catch(error =>
+            console.log("Error creating ID and Checking Out: ", error)
+          )
+      );
+
     return (
-      <CheckoutButton className="checkout" onClick={(goToShopifyCheckout)}>
+      <CheckoutButton className="checkout" onClick={createIdAndCheckout}>
```

Comment:
I'd probably do something like this. I can't spin it up and test it because i don't have keys, so there could possibly be some errors here, but this is the gist. Async / Await really helps avoid nesting, and also breaking that function createButton function up into two functions are what I would try and do. In functional programming, people try to write their functions with one clear outcome.
```
  createIdAndCheckout = async lineItemsToAdd => () => {
    const history = useHistory();
    try {
      const checkout = await client.checkout.create();
      await updateCheckoutId(checkout.id);
      await client.checkout.addLineItems(checkout.id, lineItemsToAdd);
      history.push(`${checkout.webUrl}`);
    } catch (error) {
      // Are you going to show any warnings to the user if this happens? 
      console.log("Error creating ID and Checking Out: ", error);
    }
  }
  
  createCheckoutButton = () => {
    const { lineItems, updateCheckoutId } = this.props;
    const createLineItemObject = item => ({
      variantId: item.variants.edges[0].node.id,
      quantity: item.quantity
    });
    const lineItemsToAdd = lineItems.map(createLineItemObject);
    const checkout = createIdAndCheckout(lineItemsToAdd);

    return (
      <CheckoutButton className="checkout" onClick={checkout}>
        Proceed to Checkout
      </CheckoutButton>
    );
  }

```

Reply:
This totally makes sense. I wasn't able to get useHistory(); to navigate to an external URL, and after digging into the docs and doing some googling, I couldn't find a way to use it in this way.

Thank you for your help!